Wang Haitao iP

HAITAO2003 opened the pull request:
# Duke Task Manager: Now with JavaFX GUI! 🎨

> A personal task management application that helps you keep track of todos, deadlines, and events.

## Major Updates in this PR
* Replaced command-line interface with a modern JavaFX GUI
* Added interactive chat-style interface
* Improved user experience with visual feedback
* Maintained core functionality while upgrading the UI

### Implementation Details
1. Created new JavaFX components:
   - Main application window
   - Chat dialog boxes
   - Input field with send button
   - Scrollable conversation view

2. Modified existing components:
   - Updated `Duke` class to support GUI
   - Enhanced `Ui` class for visual output
   - Maintained backward compatibility

### Code Example
```java
public class DialogBox extends HBox {
    @FXML
    private Label dialog;
    @FXML
    private ImageView displayPicture;

    private DialogBox(String text, String imagePath) {
        try {
            FXMLLoader fxmlLoader = new FXMLLoader(Main.class.getResource("/view/DialogBox.fxml"));
            fxmlLoader.setController(this);
            fxmlLoader.setRoot(this);
            fxmlLoader.load();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

### Project Status
- [x] JavaFX GUI implementation
- [x] Chat interface
- [x] Basic styling
- [ ] Advanced features
- [ ] User documentation
- [ ] Testing

## Technical Details
The project now uses `JavaFX` for the GUI, with **FXML** for layout management. The original functionality is maintained through the ~~command line interface~~ new graphical interface.

For more information about JavaFX, visit [OpenJFX documentation](https://openjfx.io/javadoc/17/).

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

HAITAO2003 committed:
Level 0 implementation finished

HAITAO2003 committed:
level 1 implementation finished

HAITAO2003 committed:
level 2 implementation completed; fixed inconsistency in code

HAITAO2003 committed:
level 3 implementation

HAITAO2003 committed:
level 4 implementation completed

HAITAO2003 committed:
text-ui-testing done

HAITAO2003 committed:
level 5 debuging done

HAITAO2003 committed:
level 7 completed.
1. Better modularisation of the code
2. added delete function
3. fixed output format problem

HAITAO2003 committed:
A-Enums

HAITAO2003 committed:
Level-7

HAITAO2003 committed:
Merge branch 'branch-Level-7'

HAITAO2003 committed:
Add Dates and Times

HAITAO2003 committed:
Fix bugs

HAITAO2003 committed:
Merge branch 'branch-Level-8'

DESU-CLUB commented on src/out/production/ip/Deadline.class:
You might want to git ignore the .class

DESU-CLUB commented on src/main/java/Task.java:
Great naming conventions here for the formats!

DESU-CLUB commented on src/main/java/Duke.java:
```diff
@@ -1,82 +1,94 @@
-                    } else {
+            try {
+                switch(parts[0]) {
+                    case "deadline":
```
Based on the coding standards, cases have no indentations (same indent as switch)

DESU-CLUB commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,68 @@
+import utils.Utils;
+
+import java.util.ArrayList;
```
Might be better to put ArrayList before custom imports like utils, as that caused me to violate the checkstyle

However the coding standards dont enforce this to that degree so this is really up to your personal preference

DESU-CLUB commented on src/main/java/Duke.java:
```diff
@@ -3,7 +3,7 @@
 import java.util.Optional;
```
I really like how you're using typing in order to annotate your classes and allow for empties!

DESU-CLUB reviewed (changes requested):
All in all well done! There were a few violations but it was a really good job with the overall naming conventions

Jiayouzz for IP!

TeaMakesMePee commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,92 @@
+        FileHandler.saveTasksToFile(this.list);
+    }
+
+    public void add_item(Task item) {
+        this.list.add(item);
+        System.out.println("Got it. I've added this task:");
+        printTask(item);
+        Utils.output("Now you have " + this.list.size() + " tasks in the list.");
+        saveToFile();
+    }
+
+    public void delete_item_by_index(int index) throws DukeException {
+        if (index < 1 || index > list.size()) {
+            throw new DukeException("Invalid task number! Please provide a number between 1 and " + list.size());
+        }
+        Task removedTask = list.remove(index - 1);
+        System.out.println("Noted. I've removed this task:");
+        printTask(removedTask);
+        Utils.output("Now you have " + this.list.size() + " tasks in the list.");
+        saveToFile();
+
+    }
+
+    public void mark_done_by_index(int idx) {
+        if (idx > this.list.size()){
+            Utils.output("Task number does not exist!");
+        }else{
+            System.out.println("Nice! I've marked this task as done:");
+            Task item = this.list.get(idx - 1);
+            item.markDone();
+            printTask(item);
+            saveToFile();
+        }
+    }
```
Please follow the naming convention for class methods. 
Suggested:
addItem(args)
deleteItem(args)
markDone(args)

HAITAO2003 committed:
A-MoreOOP

HAITAO2003 committed:
Add better package organisation

HAITAO2003 committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

HAITAO2003 committed:
Test all non-trivial method in Deadline class

HAITAO2003 committed:
Test all non-trivial methods in Deadline.java

HAITAO2003 committed:
update build.gradle

HAITAO2003 committed:
Fix coding standard violation

HAITAO2003 committed:
add find task funtionality

HAITAO2003 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fix coding standard violation

# Conflicts:
#	src/main/java/duke/Parser.java

HAITAO2003 committed:
Add Javadoc to FindCommand.java

HAITAO2003 committed:
Add GUI

HAITAO2003 committed:
Add GUI

HAITAO2003 committed:
Merge branch 'master' into branch-Level-9

* master:
  Add GUI

HAITAO2003 committed:
Revert "Merge branch 'master' into branch-Level-9"

This reverts commit 5780fd4617df58f08b77b5215a483d1e639c43a9, reversing
changes made to f83a5cdb8184c90b035e10f2d517366a5acd4db4.

HAITAO2003 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:

HAITAO2003 committed:
Initial commit

HAITAO2003 committed:
feat: add assertions to document code invariants

Add comprehensive assertions throughout Parser and Storage classes to
document and validate important code assumptions and invariants.
These assertions serve as both documentation and runtime checks
during development.

Parser changes:
- Add input validation assertions for all parameters
- Validate command string formatting and processing
- Add assertions for task-specific format requirements
- Ensure index operations remain within valid bounds
- Document state invariants after string operations

Storage changes:
- Add constructor validation for file path parameters
- Validate data structure integrity for task lists
- Add file operation invariants for read/write operations
- Ensure task conversion maintains data integrity
- Add format validation for task string parsing

HAITAO2003 committed:
Initial commit

HAITAO2003 committed:
docs: add Javadoc documentation to Parser class

Add comprehensive Javadoc comments to improve code documentation:
- Add class-level documentation for Parser
- Document all public and private methods
- Specify input formats and validation rules
- Document error conditions and exceptions

No functional changes, documentation only.

HAITAO2003 committed:
Initial commit

HAITAO2003 committed:
Merge pull request #2 from HAITAO2003/branch-A-Streams

Initial commit

HAITAO2003 committed:
Merge pull request #3 from HAITAO2003/branch-A-Assertions

Branch a assertions

HAITAO2003 committed:
Merge pull request #4 from HAITAO2003/branch-A-CodeQuality

Branch a code quality

HAITAO2003 committed:
Initial commit

HAITAO2003 committed:
Add better gui and finish up level 10 design

HAITAO2003 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add better gui and finish up level 10 design
  Initial commit

HAITAO2003 committed:
Add extension C update function

HAITAO2003 committed:
Update README.md

HAITAO2003 committed:
Update README.md

HAITAO2003 committed:
Update README.md

HAITAO2003 committed:
Update README.md

HAITAO2003 committed:
fix: handle empty deadline date after /by keyword

Parser accepts deadline command with empty date value after /by keyword.
This causes unexpected exceptions when processing the command.

Add validation check to ensure deadline date is provided.
This ensures proper error message instead of runtime exceptions.

Similar validation needed for event parameters in future commits.

HAITAO2003 committed:
Merge branch 'master' of github.com:HAITAO2003/ip

* 'master' of github.com:HAITAO2003/ip:
  Update README.md
  Update README.md
  Update README.md
  Update README.md

HAITAO2003 committed:
docs: add JavaDoc to DialogBox class

DialogBox class lacks proper documentation for its functionality and components.
This makes it difficult for new developers to understand the code purpose and usage.

Add comprehensive JavaDoc comments to class, methods, and fields.
Documentation describes the purpose of each component without changing any existing code.

Improved documentation aids in code maintenance and facilitates better understanding
of the GUI dialog system implementation.

HAITAO2003 committed:
refactor: make Event time fields final

Event class date-time fields are mutable and could potentially be modified after creation.
This violates the principle of immutability and could lead to inconsistent state.

Make startDateTime and endDateTime fields final to ensure immutability.
This guarantees that event time data remains consistent throughout the object lifecycle.

The change is minimal but improves code robustness and maintainability.

HAITAO2003 committed:
docs: add JavaDoc to MainWindow controller class

MainWindow class lacks proper documentation of its UI components and functionality.
This makes it difficult for new developers to understand the controller's purpose and behavior.

Add comprehensive JavaDoc comments to class, fields, and methods.
Documentation clarifies the role of each UI component and method without changing any functionality.

Improved documentation enhances code maintainability and provides better understanding
of the main window controller's responsibilities in the Duke application.

HAITAO2003 committed:
style: streamline JavaDoc in MainWindow class

Field-level JavaDoc comments are redundant and state the obvious.
These verbose comments add noise without providing meaningful information.

Remove excessive field-level JavaDoc comments while preserving method documentation.
This maintains meaningful documentation while improving code readability.

Class and method JavaDoc remain intact to preserve important contextual information.

HAITAO2003 committed:
refactor: use pattern matching for instanceof in formatTask

Method uses type casting after instanceof checks to access subclass methods.
This approach is verbose and requires redundant type casting.

Implement pattern matching with instanceof to simplify variable handling.

The refactoring reduces code verbosity while maintaining the same functionality.

HAITAO2003 committed:
refactor: update Storage class with Java features

Storage class uses verbose lambda expressions, manual type casting, and has redundant assertions.
These patterns make the code unnecessarily verbose and harder to maintain.

Replace lambdas with method references, implement pattern matching, and remove redundant assertions.
This leverages Java's modern features for more concise code while maintaining necessary validations.

Code functionality remains identical while reducing verbosity and potential for errors.

HAITAO2003 committed:
docs: add GUI screenshot and README to docs folder

Project lacks visual representation and documentation in standard location.
This makes it difficult for users to understand the application's appearance and usage.

Add GUI screenshot and README file to the docs directory.
This follows standard practice for GitHub documentation and improves project presentation.

Documentation in the proper location improves discoverability for new users and contributors.

HAITAO2003 committed:
Update README.md

HAITAO2003 committed:
Update README.md

fix link to release page

HAITAO2003 committed:
feat: rename mascot from Duke to Dumpling

Application's mascot name "Duke" is generic and lacks personality.
A more distinctive name would enhance the user experience and brand identity.

Change mascot name from "Duke" to "Dumpling" throughout the application.
This includes updating UI text, documentation, and related references.

The new name adds uniqueness and character to the application while maintaining
the same functionality and user interaction patterns.

HAITAO2003 committed:
fix: resolve JavaFX runtime component dependency issues

- Add Launcher class to properly initialize JavaFX application
- Update build.gradle to include platform-specific JavaFX dependencies
- Configure shadowJar plugin to package all dependencies in a single executable JAR
- Fix cross-platform compatibility issues for Windows, Mac, and Linux

Resolves issue with "JavaFX runtime components are missing" error
